File: vidyapath/server/models/ServiceProvider.js
const mongoose = require('mongoose');

const serviceProviderSchema = new mongoose.Schema({
  name: { type: String, required: true },
  tagline: { type: String, default: '' },
  description: { type: String, default: '' },
  link: { type: String, default: '' },
  linkLabel: { type: String, default: 'Visit Site' },
  image: { type: String, default: '' },
  category: {
    type: String,
    enum: ['school', 'coaching', 'tuition', 'test-prep', 'library', 'training', 'consultancy', 'other'],
    default: 'other',
  },
  servicesOffered: [{ type: String }],
  contactPhone: { type: String, default: '' },
  contactEmail: { type: String, default: '' },
  address: { type: String, default: '' },
  city: { type: String, default: '' },
  state: { type: String, default: '' },
  website: { type: String, default: '' },
  establishedYear: Number,
  discountInfo: { type: String, default: '' },
  featured: { type: Boolean, default: false },
  order: { type: Number, default: 0 },
  startDate: Date,
  endDate: Date,
  rating: { type: Number, min: 0, max: 5, default: 0 },
  isActive: { type: Boolean, default: true },
  clickCount: { type: Number, default: 0 },
}, { timestamps: true });

serviceProviderSchema.index({ isActive: 1, featured: -1, order: 1 });
serviceProviderSchema.index({ category: 1, city: 1, isActive: 1 });

module.exports = mongoose.model('ServiceProvider', serviceProviderSchema);

File: vidyapath_app/lib/models/service_provider.dart
class ServiceProviderModel {
  final String id;
  final String name;
  final String? tagline;
  final String? description;
  final String? link;
  final String? linkLabel;
  final String? image;
  final String? category;
  final List<String> servicesOffered;
  final String? contactPhone;
  final String? contactEmail;
  final String? city;
  final String? state;
  final String? website;
  final int? establishedYear;
  final String? discountInfo;
  final bool featured;
  final int clickCount;
  final double rating;
  final DateTime? createdAt;

  ServiceProviderModel({
    required this.id,
    required this.name,
    this.tagline,
    this.description,
    this.link,
    this.linkLabel,
    this.image,
    this.category,
    this.servicesOffered = const [],
    this.contactPhone,
    this.contactEmail,
    this.city,
    this.state,
    this.website,
    this.establishedYear,
    this.discountInfo,
    this.featured = false,
    this.clickCount = 0,
    this.rating = 0,
    this.createdAt,
  });

  factory ServiceProviderModel.fromJson(Map<String, dynamic> json) {
    return ServiceProviderModel(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      tagline: json['tagline'],
      description: json['description'],
      link: json['link'],
      linkLabel: json['linkLabel'],
      image: json['image'],
      category: json['category'],
      servicesOffered: json['servicesOffered'] is List ? List<String>.from(json['servicesOffered']) : [],
      contactPhone: json['contactPhone'],
      contactEmail: json['contactEmail'],
      city: json['city'],
      state: json['state'],
      website: json['website'],
      establishedYear: json['establishedYear'],
      discountInfo: json['discountInfo'],
      featured: json['featured'] ?? false,
      clickCount: json['clickCount'] ?? 0,
      rating: (json['rating'] as num?)?.toDouble() ?? 0,
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt']) : null,
    );
  }

  String get location => [city, state].where((s) => s != null && s.isNotEmpty).join(', ');
}
